File: xingchen/code_module_scoring.js
const fs = require("fs");
const path = require("path");

function readJson(filePath) {
  return JSON.parse(fs.readFileSync(filePath, "utf8"));
}

function loadDefaultData() {
  const root = path.resolve(__dirname, "..");
  return {
    questions: readJson(path.join(root, "diagnosis", "diagnostic_questions.json")),
    rules: readJson(path.join(root, "diagnosis", "scoring_rules.json"))
  };
}

function shortQuestionId(questionId) {
  const match = String(questionId).match(/^Q0*(\d+)$/i);
  return match ? `Q${String(Number(match[1])).padStart(2, "0")}` : String(questionId);
}

function longQuestionId(questionId) {
  const match = String(questionId).match(/^Q0*(\d+)$/i);
  return match ? `Q${String(Number(match[1])).padStart(3, "0")}` : String(questionId);
}

function normalizeChoice(value) {
  return String(value || "").trim().toUpperCase();
}

function normalizeChoiceList(value) {
  if (Array.isArray(value)) {
    return value.map(normalizeChoice).filter(Boolean);
  }

  const raw = String(value || "").trim().toUpperCase();
  if (!raw) {
    return [];
  }

  if (/[,，\s|>]+/.test(raw)) {
    return raw.split(/[,，\s|>]+/).map(normalizeChoice).filter(Boolean);
  }

  if (/^[A-Z]+$/.test(raw) && raw.length > 1) {
    return raw.split("");
  }

  return [raw];
}

function sameSet(left, right) {
  const a = normalizeChoiceList(left).sort();
  const b = normalizeChoiceList(right).sort();
  return a.length === b.length && a.every((value, index) => value === b[index]);
}

function sameOrder(left, right) {
  const a = normalizeChoiceList(left);
  const b = normalizeChoiceList(right);
  return a.length === b.length && a.every((value, index) => value === b[index]);
}

function isCorrect(answer, question) {
  if (answer === undefined || answer === null || answer === "") {
    return false;
  }

  if (question.type === "multiple_choice") {
    return sameSet(answer, question.correct_answer);
  }

  if (question.type === "ordering") {
    return sameOrder(answer, question.correct_answer);
  }

  return normalizeChoice(answer) === normalizeChoice(question.correct_answer);
}

function buildQuestionMaps(questionData) {
  const byLongId = new Map();
  const byShortId = new Map();

  for (const question of questionData.questions || []) {
    byLongId.set(question.id, question);
    byShortId.set(shortQuestionId(question.id), question);
  }

  return { byLongId, byShortId };
}

function questionRuleFor(rules, questionId) {
  const questionRules = rules.questions || {};
  return questionRules[questionId] || questionRules[longQuestionId(questionId)] || questionRules[shortQuestionId(questionId)] || {};
}

function addUnique(list, seen, value) {
  if (value && !seen.has(value)) {
    seen.add(value);
    list.push(value);
  }
}

function addWeakAbility(output, seen, abilityKey, reason, rules) {
  const catalog = (rules.ability_catalog || {})[abilityKey] || {};
  const abilityId = catalog.ability_id || abilityKey;

  if (seen.has(abilityId)) {
    return;
  }

  seen.add(abilityId);
  output.push({
    ability_id: abilityId,
    ability_name: catalog.ability_name || abilityKey,
    reason: reason || catalog.default_reason || "该能力点对应诊断题回答错误"
  });
}

function feedbackLevel(score, weakAbilities, rules) {
  const critical = new Set(rules.critical_abilities || []);
  const hasCriticalWeakness = weakAbilities.some((item) => critical.has(item.ability_id));

  if (score < 100 && hasCriticalWeakness) {
    return "需要补基础";
  }

  const levels = (rules.feedback_levels || []).slice().sort((a, b) => b.min_score - a.min_score);
  const matched = levels.find((level) => score >= level.min_score);
  return matched ? matched.label : "需要补基础";
}

function scoreDiagnostic(input, data) {
  const loaded = data || loadDefaultData();
  const questionData = loaded.questions;
  const rules = loaded.rules;
  const answers = (input && input.answers) || {};
  const { byShortId } = buildQuestionMaps(questionData);
  const weakAbilities = [];
  const weakSeen = new Set();
  const recommendedPath = [];
  const pathSeen = new Set();
  let correctCount = 0;

  for (const [questionId, question] of byShortId.entries()) {
    const answer = answers[questionId] !== undefined ? answers[questionId] : answers[question.id];
    const correct = isCorrect(answer, question);

    if (correct) {
      correctCount += 1;
      continue;
    }

    const rule = questionRuleFor(rules, question.id);
    const reason = rule.reason || question.wrong_feedback;

    for (const abilityKey of rule.weak_abilities || [question.ability_id]) {
      addWeakAbility(weakAbilities, weakSeen, abilityKey, reason, rules);
    }

    for (const item of rule.recommended_path || []) {
      addUnique(recommendedPath, pathSeen, item);
    }
  }

  const totalCount = byShortId.size;
  const score = totalCount ? Math.round((correctCount / totalCount) * 100) : 0;

  return {
    score,
    correct_count: correctCount,
    total_count: totalCount,
    weak_abilities: weakAbilities,
    recommended_path: recommendedPath,
    feedback_level: feedbackLevel(score, weakAbilities, rules)
  };
}

function main(input) {
  return scoreDiagnostic(input);
}

if (require.main === module) {
  const raw = fs.readFileSync(0, "utf8").trim();
  const input = raw ? JSON.parse(raw) : { answers: {} };
  process.stdout.write(JSON.stringify(main(input), null, 2));
}

module.exports = {
  main,
  scoreDiagnostic
};

File: xingchen/code_module_scoring.py
import json
import re
import sys
from pathlib import Path


def read_json(file_path):
    return json.loads(Path(file_path).read_text(encoding="utf-8"))


def load_default_data():
    root = Path(__file__).resolve().parent.parent
    return {
        "questions": read_json(root / "diagnosis" / "diagnostic_questions.json"),
        "rules": read_json(root / "diagnosis" / "scoring_rules.json"),
    }


def short_question_id(question_id):
    match = re.match(r"^Q0*(\d+)$", str(question_id), re.I)
    return f"Q{int(match.group(1)):02d}" if match else str(question_id)


def long_question_id(question_id):
    match = re.match(r"^Q0*(\d+)$", str(question_id), re.I)
    return f"Q{int(match.group(1)):03d}" if match else str(question_id)


def normalize_choice(value):
    return str(value or "").strip().upper()


def normalize_choice_list(value):
    if isinstance(value, list):
        return [normalize_choice(item) for item in value if normalize_choice(item)]

    raw = normalize_choice(value)
    if not raw:
        return []

    if re.search(r"[,，\s|>]+", raw):
        return [normalize_choice(item) for item in re.split(r"[,，\s|>]+", raw) if normalize_choice(item)]

    if re.match(r"^[A-Z]+$", raw) and len(raw) > 1:
        return list(raw)

    return [raw]


def same_set(left, right):
    return sorted(normalize_choice_list(left)) == sorted(normalize_choice_list(right))


def same_order(left, right):
    return normalize_choice_list(left) == normalize_choice_list(right)


def is_correct(answer, question):
    if answer is None or answer == "":
        return False

    if question.get("type") == "multiple_choice":
        return same_set(answer, question.get("correct_answer", []))

    if question.get("type") == "ordering":
        return same_order(answer, question.get("correct_answer", []))

    return normalize_choice(answer) == normalize_choice(question.get("correct_answer"))


def build_question_maps(question_data):
    by_short_id = {}
    for question in question_data.get("questions", []):
        by_short_id[short_question_id(question["id"])] = question
    return by_short_id


def question_rule_for(rules, question_id):
    question_rules = rules.get("questions", {})
    return (
        question_rules.get(question_id)
        or question_rules.get(long_question_id(question_id))
        or question_rules.get(short_question_id(question_id))
        or {}
    )


def add_unique(items, seen, value):
    if value and value not in seen:
        seen.add(value)
        items.append(value)


def add_weak_ability(output, seen, ability_key, reason, rules):
    catalog = rules.get("ability_catalog", {}).get(ability_key, {})
    ability_id = catalog.get("ability_id", ability_key)

    if ability_id in seen:
        return

    seen.add(ability_id)
    output.append(
        {
            "ability_id": ability_id,
            "ability_name": catalog.get("ability_name", ability_key),
            "reason": reason or catalog.get("default_reason") or "该能力点对应诊断题回答错误",
        }
    )


def feedback_level(score, weak_abilities, rules):
    critical = set(rules.get("critical_abilities", []))
    has_critical_weakness = any(item.get("ability_id") in critical for item in weak_abilities)

    if score < 100 and has_critical_weakness:
        return "需要补基础"

    levels = sorted(rules.get("feedback_levels", []), key=lambda item: item.get("min_score", 0), reverse=True)
    for level in levels:
        if score >= level.get("min_score", 0):
            return level.get("label", "需要补基础")
    return "需要补基础"


def score_diagnostic(input_data, data=None):
    loaded = data or load_default_data()
    question_data = loaded["questions"]
    rules = loaded["rules"]
    answers = (input_data or {}).get("answers", {})
    questions = build_question_maps(question_data)
    weak_abilities = []
    weak_seen = set()
    recommended_path = []
    path_seen = set()
    correct_count = 0

    for question_id, question in questions.items():
        answer = answers.get(question_id, answers.get(question["id"]))

        if is_correct(answer, question):
            correct_count += 1
            continue

        rule = question_rule_for(rules, question["id"])
        reason = rule.get("reason") or question.get("wrong_feedback")

        for ability_key in rule.get("weak_abilities", [question.get("ability_id")]):
            add_weak_ability(weak_abilities, weak_seen, ability_key, reason, rules)

        for item in rule.get("recommended_path", []):
            add_unique(recommended_path, path_seen, item)

    total_count = len(questions)
    score = round((correct_count / total_count) * 100) if total_count else 0

    return {
        "score": score,
        "correct_count": correct_count,
        "total_count": total_count,
        "weak_abilities": weak_abilities,
        "recommended_path": recommended_path,
        "feedback_level": feedback_level(score, weak_abilities, rules),
    }


def main(input_data):
    return score_diagnostic(input_data)


if __name__ == "__main__":
    raw = sys.stdin.read().strip()
    payload = json.loads(raw) if raw else {"answers": {}}
    print(json.dumps(main(payload), ensure_ascii=False, indent=2))
